portal-impl/src/com/liferay/portlet/internal/PortletBagImpl.java

```diff
@@ -184,16 +184,21 @@ public List<PreferencesValidator> getPreferencesValidatorInstances() {
 
 	@Override
 	public ResourceBundle getResourceBundle(Locale locale) {
+		Snapshot<ResourceBundleLoader> resourceBundleLoaderSnapshot =
```

Comment:
@JanisZhang This is not right.

Before change, the `ServiceProxyFactory` usage is in a `DCLSingleton`. The `DCLSingleton` here is to make it lazy and singleton. But the `ServiceProxyFactory` can still replace the `_singleton` in `DCLSingleton` if a real service is available.

The equivalent should be a dynamic `Snapshot` instance field. And we may not need the dummy service, because this method doesn't expose the service; it returns something obtained from the service. So, if service is null, we can shortcut.

This way, a service tracker is replaced by another service tracker, while we reduce the proxy creation. The thing is still lazy.